vermouth/ffinput.py

```diff
@@ -51,196 +55,337 @@
-                if not brackets and end != start:
-                    end -= 1
-                    break
-                brackets += 1
-            elif end_char == '}':
-                brackets -= 1
-                if not brackets:
-                    break
-            elif end_char in separators:
-                if not brackets:
-                    # We reached a separator. We do not want the separator to
-                    # be included in the token, so we push the end by one
-                    # character to the left.
-                    end -= 1
-                    break
-        if brackets > 0:
-            msg = 'Unexpected end of line. A closing bracket is missing.'
-            raise IOError(msg)
-        elif brackets < 0:
-            msg = 'An opening bracket is missing.'
-            raise IOError(msg)
-
-        token = line[start:end + 1]
-        if token:
-            tokens.append(token)
-
-        # Find the beginning of the next token.
-        start = end + 1
-        while start < len(line) and line[start] in separators:
-            start += 1
-    return tokens
-
-
-def _substitute_macros(line, macros):
-    r"""
-    Substitute macros by their content.
-
-    A macro starts with a '$' and ends with one amongst ' ${}\n\t"'.
-
-    Parameters
-    ----------
-    line: str
-        The line to fix.
-    macros: dict[str, str]
-        Keys are macro names, values are the replacement content.
+class FFDirector(SectionLineParser):
+    COMMENT_CHAR = ';'
+    interactions_natoms = {
+        'bonds': 2,
+        'angles': 3,
+        'dihedrals': 4,
+        'impropers': 4,
+        'constraints': 2,
+        'virtual_sites2': 3,
+        'pairs': 2,
+    }
 
-    Returns
-    -------
-    str
-    """
-    start = None
-    while True:  # stops when start < 0
-        start = line.find('$', start)
-        if start < 0:
-            break
-        for end, char in enumerate(line[start + 1:], start=start + 1):
-            if char in ' \t\n{}$"':
-                break
-        else: # no break
-            end += 1
-        macro_name = line[start + 1:end]
-        macro_value = macros[macro_name]
-        line = line[:start] + macro_value + line[end:]
-        end = start + len(macro_value)
-    return line
+    def __init__(self, force_field):
+        super().__init__()
+        self.force_field = force_field
+        self.current_block = None
+        self.current_link = None
+        self.current_modification = None
+        self.blocks = collections.OrderedDict()
+        self.links = []
+        self.modifications = []
+
+        self.header_actions = {
+            ('moleculetype', ): self._new_block,
+            ('link', ): self._new_link,
+            ('modification', ): self._new_modification,
+        }
+
+    def parse_header(self, line, lineno=0):
+        """
+        Parses a section header with line number `lineno`. Sets :attr:`section`
```

Comment:
The sphinx problem is that the class has no 'section' mentioned as one of it's Attributes in it's docstring (even though SectionLineParser does). Try changing the reference to ``:attr:`~SectionLineParser.section` ``. (or `vermouth.parser_utils.SectionLineParser.section`). Or something along those lines. I'll do a full review of your code asap.

Reply:
done